src/Analyzers/CSharp/Analyzers/UseSystemThreadingLock/CSharpUseSystemThreadingLockDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,235 @@
+
+        var namedType = (INamedTypeSymbol)context.Symbol;
+        if (namedType is not
+            {
+                TypeKind: TypeKind.Class or TypeKind.Struct,
+                DeclaringSyntaxReferences: [var reference, ..]
+            })
+        {
+            return;
+        }
+
+        var syntaxTree = reference.GetSyntax(cancellationToken).SyntaxTree;
+        var option = context.GetCSharpAnalyzerOptions(syntaxTree).PreferSystemThreadingLock;
+
+        // Bail immediately if the user has disabled this feature.
+        if (!option.Value || ShouldSkipAnalysis(syntaxTree, context.Options, context.Compilation.Options, option.Notification, cancellationToken))
+            return;
+
+        // Needs to have a private field that is exactly typed as 'object'.  This way we can analyze all usages of it to
+        // be sure it's completely safe to move to the new lock type.
+        using var fieldsArray = TemporaryArray<IFieldSymbol>.Empty;
+        using var _ = PooledHashSet<SemanticModel>.GetInstance(out var cachedSemanticModels);
+
+        foreach (var member in namedType.GetMembers())
+        {
+            if (member is not IFieldSymbol
+                {
+                    Type.SpecialType: SpecialType.System_Object,
+                    DeclaredAccessibility: Accessibility.Private,
+                    DeclaringSyntaxReferences: [var fieldSyntaxReference],
+                } field)
+            {
+                continue;
+            }
+
+            if (fieldSyntaxReference.GetSyntax(cancellationToken) is not VariableDeclaratorSyntax fieldSyntax)
+                continue;
+
+            // For simplicity, only offer this for fields with a single declarator.
+            if (fieldSyntax.Parent is not VariableDeclarationSyntax { Parent: FieldDeclarationSyntax, Variables.Count: 1 })
+                return;
+
+            // If we have an initializer at the declaration site, it needs to be initialized with either `new object()`
+            // or `new()`.
+            if (fieldSyntax.Initializer != null && !IsSystemObjectCreationExpression(fieldSyntax.Initializer.Value))
+                continue;
+
+            // Looks like something that could be converted to a System.Threading.Lock if we see that this field is used
+            // in a compatible fashion.
+            fieldsArray.Add(field);
+        }
+
+        if (fieldsArray.Count == 0)
+            return;
+
+        // The set of fields we think could be converted to `System.Threading.Lock` from `object`.
+        var potentialLockFields = new ConcurrentSet<IFieldSymbol>();
+
+        // Whether or not we saw this field used in a `lock (obj)` statement.  If not, we do not want to convert this as
+        // the user wasn't using this as a lock.  Note: we can consider expanding the set of patterns we detect (like
+        // Monitor.Enter + Monitor.Exit) if we think it's worthwhile.
+        var wasLockedSet = new ConcurrentSet<IFieldSymbol>();
```

Comment:
💡 This can be fairly easily combined into `potentialLockFields` if you use the value ref approach, e.g. using a value tuple.

Reply:
yup.  doing just that.